citar.el

```diff
@@ -1107,89 +1092,78 @@ For use with `embark-act-all'."
-             (citar--select-resource files)))))
-    (if file
-        (funcall fn file)
-      (message "No associated file"))))
+;; TODO Rename? This also opens files in bib field, not just library files
+;;;###autoload
+(defun citar-open-files (key-or-keys)
+  "Open library file associated with KEY-OR-KEYS."
+  (interactive (list (citar-select-refs)))
+  ;; TODO filter to refs have files?
+  (let ((embark-default-action-overrides '((file . citar-file-open))))
+    (citar--library-file-action key-or-keys #'citar-file-open)))
 
 ;;;###autoload
-(defun citar-open-library-file (key-entry)
-  "Open library file associated with the KEY-ENTRY.
+(defun citar-attach-files (key-or-keys)
+  "Attach library file associated with KEY-OR-KEYS to outgoing MIME message.
 
 With prefix, rebuild the cache before offering candidates."
-  (interactive (list (citar-select-ref
-                      :rebuild-cache current-prefix-arg)))
-  (let ((embark-default-action-overrides '((file . citar-file-open))))
-    (when (and citar-library-paths
-               (stringp citar-library-paths))
-      (error "Make sure 'citar-library-paths' is a list of paths"))
-    (citar--library-file-action key-entry 'open)))
+  (interactive (list (citar-select-ref)))
+  (let ((embark-default-action-overrides '((file . mml-attach-file))))
+    (citar--library-file-action key-or-keys #'mml-attach-file)))
+
+(defun citar--library-file-action (key-or-keys action)
+  "Run ACTION on file associated with KEY-OR-KEYS."
+  (let ((entries (citar-get-entries)))
+    (if-let ((files (citar-get-files key-or-keys :entries entries)))
+        (funcall action (if (null (cdr files))
+                            (car files)
+                          (citar--select-resource files)))
+      (ignore
+       ;; If some key had files according to `citar-has-files', but `citar-get-files' returned nothing, then
+       ;; don't print the following message. The appropriate function in `citar-get-files-functions' is
+       ;; responsible for telling the user why it failed, and we want that explanation to appear in the echo
+       ;; area.
+       (let ((keys (if (listp key-or-keys) key-or-keys (list key-or-keys)))
+             (hasfilep (citar-has-files :entries entries)))
+         (unless (and hasfilep (seq-some hasfilep keys))
+           (message "No associated files for %s" key-or-keys)))))))
 
 ;;;###autoload
-(defun citar-open-notes (key-entry)
-  "Open notes associated with the KEY-ENTRY.
-With prefix, rebuild the cache before offering candidates."
-  (interactive (list (citar-select-ref
-                      :rebuild-cache current-prefix-arg)))
-  (let* ((embark-default-action-overrides '((file . find-file)))
-         (key (car key-entry))
-         (entry (cdr key-entry)))
+(defun citar-open-notes (key)
+  "Open notes associated with the KEY."
+  ;; REVIEW KEY, or KEYS
+  (interactive (list (citar-select-ref)))
+  (let ((embark-default-action-overrides '((file . find-file))))
```

Comment:
@roshanshariff WDYT about this line?

Would it not be the case that (again, just for practical demonstration) `citar-org-roam` candidates would not be files, but org-roam "nodes"?

So would to deal with that?